ntex-bytes/src/buf/pool.rs

```diff
@@ -0,0 +1,269 @@
+#[derive(Clone, Copy, Debug)]
+pub struct BytesPool;
+
+impl BufferSource for BytesPool {
+    type Owned = BytesMut;
+    type Accumulator = BytesAccumulator;
+    type TakeFuture = future::Ready<Self::Owned>;
+
+    fn take(&self, len: usize) -> Self::TakeFuture {
+        future::ready(BytesMut::with_capacity(len))
+    }
+}
+
+impl BufferSink for BytesPool {
+    type Backing = ();
+    type Owned = <Self as BufferSource>::Owned;
+
+    fn put_back(&self, _backing: Self::Backing) {
+    }
+}
+
+/// A buffer that owns a particular range of its backing buffer.
+///
+/// An `Owned` tracks what part of itself has been filled with data.
+/// The filled region can be accessed with [`Owned::filled`].
+/// The unfilled region can be accessed with [`Owned::unfilled_mut`], and bytes can be moved from the start of this region
+/// into the end of the filled region with [`Owned::fill`].
+///
+/// An `Owned` can be subdivided into smaller `Owned`s with `split_to` that each own
+/// smaller splits of the backing buffer.
+///
+/// An `Owned` is not `Clone`. It can be converted to a [`Shared`] which is, via [`Owned::freeze`]
+pub trait Owned {
+    /// The type of shared buffers with the same backing buffer.
+    type Shared: Shared;
+
+    /// Try to append a `u8` to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u8`.
+    fn try_put_u8(&mut self, n: u8) -> Option<()> {
+        self.try_put_slice(&[n])
+    }
+
+    /// Try to append a `u16` in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u16`.
+    fn try_put_u16_be(&mut self, n: u16) -> Option<()> {
+        self.try_put_slice(&n.to_be_bytes())
+    }
+
+    /// Try to append a `u32` in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u32`.
+    fn try_put_u32_be(&mut self, n: u32) -> Option<()> {
+        self.try_put_slice(&n.to_be_bytes())
+    }
+
+    /// Try to append a slice in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this slice.
+    fn try_put_slice(&mut self, src: &[u8]) -> Option<()>;
+
+    /// Returns the filled region of this buffer.
+    fn filled(&self) -> &[u8];
+
+    /// Returns whether the filled region of this buffer is empty or not.
+    fn filled_is_empty(&self) -> bool {
+        self.filled().is_empty()
+    }
+
+    /// Retains the range `i..` in `self`, and returns a new `Owned` for the range `0..i`
+    fn split_to(&mut self, i: usize) -> Self;
+
+    /// Returns the unfilled region of this buffer.
+    fn unfilled_mut(&mut self) -> &mut UninitSlice;
+
+    /// Moves the given number of bytes from the start of the unfilled region to the end of the filled region.
+    unsafe fn fill(&mut self, n: usize);
```

Comment:
for api consistency should it be `advance`?

Reply:
In [the original iotedge-broker code,](https://github.com/Azure/iotedge-broker/blob/46147f14c2dcc775bc321ca169ac67b1d6f20d42/data-plane/buffer-pool/src/lib.rs#L151-L163) we have one operation for advancing the end-of-filled-region cursor (called by writer) and one for advancing the beginning-of-filled-region cursor (called by reader), so we called it `fill` and `drain`.

I haven't added `drain` yet to this PR because I'm adding functions to the traits from the original types as I get to their call sites. It will be there eventually.

`bytes` calls them `advance_mut` and `advance` respectively, which I do not find descriptive and would like to avoid.

Reply:
bytes names are pretty descriptive, advance just changes internal attr as opposite to fill which does the same but does not modify any data.

Reply:
I'm open to changing the names to something that unambiguously makes it clear that one changes the end position of the filled region and the other changes the start position of the filled region. I understand that `fill` and `drain` don't make that clear, but I also don't think `advance_mut` and `advance` are clear.

Reply:
`advance` and `advance_mut` are here for some time, i've never seen `fill` being use for what it does